cumulusci/tasks/datadictionary.py

```diff
@@ -189,102 +386,160 @@ def _process_field_element(self, sobject_name, field, version):
 
                     valid_values = "; ".join(names)
             elif field_type == "Lookup":
-                valid_values = "->" + field.referenceTo.text
-            else:
-                valid_values = ""
-
-            self._set_version_with_props(
-                self.schema[sobject_name]["fields"][field_name],
-                {
-                    "version": version,
-                    "help_text": help_text_elem.text
-                    if help_text_elem is not None
-                    else "",
-                    "description": description_text_elem.text
-                    if description_text_elem is not None
-                    else "",
-                    "label": field.label.text,
-                    "valid_values": valid_values,
-                    "type": field_type,
-                },
+                target_sobject = field.referenceTo.text
+                if target_sobject.count("__") == 1:
+                    target_sobject = f"{version.package.namespace}{target_sobject}"
+                field_type = f"Lookup to {target_sobject}"
+                # Note: polymorphic custom fields are not allowed.
+            elif field_type == "MasterDetail":
+                target_sobject = field.referenceTo.text
+                if target_sobject.count("__") == 1:
+                    target_sobject = f"{version.package.namespace}{target_sobject}"
+                field_type = f"Master-Detail Relationship to {target_sobject}"
+                # Note: polymorphic custom fields are not allowed.
+
+            fd = FieldDetail(
+                version,
+                sobject,
+                sobject_label,
+                field_name,
+                field.label.text,
+                f"{field_type}{length}",
+                help_text_elem.text if help_text_elem else "",
+                description_text_elem.text if description_text_elem else "",
+                valid_values,
             )
+            fully_qualified_name = f"{sobject}.{fd.api_name}"
+            self.fields[fully_qualified_name].append(fd)
+
+    def _write_object_results(self, file_handle):
+        """Write to the given handle an output CSV containing the data dictionary for sObjects."""
+        writer = csv.writer(file_handle)
+
+        writer.writerow(
+            [
+                "Object Label",
+                "Object API Name",
+                "Object Description",
+                "Version Introduced",
+                "Version Deleted",
+            ]
+        )
 
-    def _write_results(self):
-        """Write the stored schema details to our destination CSVs"""
-        with open(self.options["object_path"], "w") as object_file:
-            writer = csv.writer(object_file)
+        for sobject_name, versions in self.sobjects.items():
+            versions = sorted(
+                versions, key=lambda ver: ver.version.version, reverse=True
```

Comment:
I got stuck here for a bit deciphering the different kinds of versions. Also, why reverse?

Reply:
An older iteration of the algorithm for finding the right versions for each column required walking the version list in reverse-chronological order. That's no longer needed, so I'll remove it.

I'll add comments to clarify the nature of the data model here.

Reply:
Scratch that, it's actually still required!